package/nwipe: switch from static tarball to Git commit (051e1aa)

Knogle opened the pull request:
This commit updates the nwipe package to build from a specific Git commit instead of using the previously referenced tarball release. The changes include:

- Setting NWIPE_SITE to the GitHub repository and NWIPE_SITE_METHOD to git.
- Specifying the commit hash (051e1aa) in NWIPE_VERSION.
- Removing or emptying the nwipe.hash file, as we no longer verify a tarball and instead rely on the Git checkout.

This approach allows ShredOS/Buildroot to clone the nwipe repository at commit 051e1aa, apply the banner patch, and build the nwipe binary directly from Git.

You can specify the desired commit in the nwipe.mk file.

PartialVolume commented:
Thanks, Looks good, it will save us some time messing about calculating hashes.

I'll give it a try over the next couple of days and then commit.

DaanSelen commented:
I was unable to recreate this scenario reliably. This was all due to the hash generation which I could not reproduce.

desertwitch commented:
For what it's worth, I've been able to build using this method and it worked fine across multiple builds, although using the full SHA-1 which the Buildroot manual also seems to recommend for this method. Here's how I used it:

```
################################################################################
#
# nwipe
#
################################################################################

# Instead of v0.38, we now use the desired commit (e.g., 051e1aa)
NWIPE_VERSION = c505c1a3a65093be629a5862ded7bf3c5506566d

# Specify the Git source:
#  - NWIPE_SITE points to the repository
#  - NWIPE_SITE_METHOD = git ensures that Buildroot performs a Git checkout
NWIPE_SITE = https://github.com/desertwitch/nwipe-pc.git
NWIPE_SITE_METHOD = git

# Dependencies remain the same (adjust them if you need additional ones)
NWIPE_DEPENDENCIES = ncurses parted dmidecode coreutils

# This hook copies your patch script (banner patch) into the source directory
# and runs autogen.sh before the actual configure/make starts.
define NWIPE_INITSH
        (cd $(@D) && cp ../../../package/nwipe/002-nwipe-banner-patch.sh $(@D) \
                && ./002-nwipe-banner-patch.sh \
                && PATH="../../host/bin:${PATH}" ./autogen.sh);
endef

NWIPE_PRE_CONFIGURE_HOOKS += NWIPE_INITSH

# Include the Buildroot autotools package framework,
# which takes care of configure/make/make install.
$(eval $(autotools-package))
```

Knogle commented:
Works for me too!

desertwitch commented on package/nwipe/nwipe.mk:
```diff
@@ -4,15 +4,20 @@
-NWIPE_VERSION = v0.38
-NWIPE_SITE = $(call github,martijnvanbrummelen,nwipe,$(NWIPE_VERSION))
-NWIPE_DEPENDENCIES = ncurses parted dmidecode coreutils
+NWIPE_VERSION = $(call qstrip,$(BR2_PACKAGE_NWIPE_GIT_REF))
```
Are you sure this will still exist if the dependency `depends on` of custom is not met? Also reset correctly back to the right default if the user was on custom before (with a value of their choice) and then switches back to e.g. default, or edited the configuration manually (in the file)?

Knogle replied:
We can also specify the repo in the menuconfig itself, so a `make defconfig` or `make clean` will solve this case.
What's your opinion on that? I can implement it quite quickly.

desertwitch replied:
I like the Kconfig variant you have here a lot, but I think the less brittle would be:
- keep the `choice` with one bool per option (as you have it)
- introduce a separate string for each option (outside the `choice`)
   - with an appropriate `default`, and `depends on` linking each respective bool (which will toggle visibility)

In the `.mk` file just a check against the bools with `ifeq, else ifeq, else ifeq` - setting the version variable.

It'd be more verbose than your solution, but I think worthwhile because then it'd check against the bool, and if that isn't set even a leftover string value from manual configuration edits or odd configurational states wouldn't break anything. Strings that don't have a corresponding bool set just wouldn't get accessed (so we don't care if they exist or not). And then it'd also not rely on `default` resetting variables when the bool isn't set, which I'm not sure it'll do.

desertwitch replied:
P.S Well, the options which only need that default (a fixed value) and should not be configurable don't need a separate string, but essentially I'd check against the bools with `ifeq` (and retrieve a separate string variable where necessary, if it's something user-settable) but not rely on Kconfig setting one string to some value itself (and mix fixed values with configurable values, based on `default` or whatever).

So something like this:
```
if BR2_PACKAGE_NWIPE

choice
	prompt "nwipe Version"
	default BR2_PACKAGE_NWIPE_VERSION_STABLE

config BR2_PACKAGE_NWIPE_VERSION_STABLE
	bool "Stable commit"

config BR2_PACKAGE_NWIPE_VERSION_MASTER
	bool "Master branch"

config BR2_PACKAGE_NWIPE_VERSION_CUSTOM
	bool "Custom Git-ref"

endchoice

config BR2_PACKAGE_NWIPE_GIT_REF
	string "Custom nwipe Git-Ref (Tag/Release/Commit)"
	depends on BR2_PACKAGE_NWIPE_VERSION_CUSTOM
	help
	  Specify custom Git-Ref:
	    - Tag/Release: e.g. "v0.38"
	    - Commit: e.g. "051e1aa"

endif
```

and then

```
ifeq ($(BR2_PACKAGE_NWIPE_VERSION_STABLE),y)
NWIPE_VERSION = 051e1aa
else ifeq ($(BR2_PACKAGE_NWIPE_VERSION_MASTER),y)
NWIPE_VERSION = master
else ifeq ($(BR2_PACKAGE_NWIPE_VERSION_CUSTOM),y)
NWIPE_VERSION = $(call qstrip,$(BR2_PACKAGE_NWIPE_GIT_REF))
endif
```

~Maybe worth dropping the `stable commit` altogether and just do `master branch` (bool) and `custom git-ref` (bool + string). That would allow for both the configuration and `.mk` file to be static, but still user-configurable.~

P.P.S: **Also beware, you cannot use branch names for `*_VERSION`** (`Note: Using a branch name as FOO_VERSION is not supported, because it does not and can not work as people would expect it should` - @ Buildroot manual). Not sure which options would be useful given that limitation... or if even the Kconfig at all.

Knogle commented:
Now we are using the 'coreboot' way to deal with this :)
@desertwitch what do you think about this one?

<img width="1220" height="962" alt="Screenshot From 2025-11-24 12-51-21" src="https://github.com/user-attachments/assets/1f73855a-dfc1-409c-a5cc-25275f58e285" />


<img width="1044" height="271" alt="Screenshot From 2025-11-24 13-22-54" src="https://github.com/user-attachments/assets/bc0e853b-90b9-43c9-a298-fecfe0e9efb3" />

desertwitch commented on package/nwipe/nwipe.mk:
```diff
@@ -4,20 +4,47 @@
+ifeq ($(BR2_PACKAGE_NWIPE_VERSION_STABLE),y)
+NWIPE_VERSION = 051e1aa
+else ifeq ($(BR2_PACKAGE_NWIPE_VERSION_MASTER),y)
+NWIPE_VERSION = master
```
According to the Buildroot manual, a branch is not possible for a package `*_VERSION`, so we'd have to test if this really works - despite `coreboot` seemingly using this already - I haven't tested it yet.

desertwitch commented on package/nwipe/nwipe.mk:
```diff
@@ -4,20 +4,47 @@
+
+# Default Git repository URL. This ensures NWIPE_SITE is never empty,
+# even if BR2_PACKAGE_NWIPE_SITE is not set in the configuration.
+NWIPE_SITE = https://github.com/martijnvanbrummelen/nwipe.git
```
Unset variables (not existing in the configuration) should assume their defaults, but I guess it won't hurt as a sanity-check either... unless something "waits" for any value to appear and immediately takes that (don't think so though).

desertwitch commented:
> Now we are using the 'coreboot' way to deal with this :) @desertwitch what do you think about this one?

I think this looks much better and will be less brittle, only not sure about the branch:

<img width="1753" height="270" alt="grafik" src="https://github.com/user-attachments/assets/4041e5cf-db00-4609-9b60-c022a8cfdeb1" />

Knogle commented:
> > Now we are using the 'coreboot' way to deal with this :) @desertwitch what do you think about this one?
> 
> I think this looks much better and will be less brittle, only not sure about the branch:
> <img alt="grafik" width="1753" height="270" src="https://private-user-images.githubusercontent.com/24509509/518110005-4041e5cf-db00-4609-9b60-c022a8cfdeb1.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjM5ODg4NzgsIm5iZiI6MTc2Mzk4ODU3OCwicGF0aCI6Ii8yNDUwOTUwOS81MTgxMTAwMDUtNDA0MWU1Y2YtZGIwMC00NjA5LTliNjAtYzAyMmE4Y2ZkZWIxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTI0VDEyNDkzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ3ZjkyZGY1YzE3MTk2ZjJmNTg4YTdjZWQ3MWYxYmQyNDMzNjBmNzE5MTJhMDc3NmM2NGFjYmRjMWVmNjBlM2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.csWYrJTY1MhpvIWlw6jOYIazwuyHUd-NOE2w2h7slw4">

Can you maybe give it some tries? For my tests it worked fine so far

PartialVolume commented:
Does this need further testing/work or are you happy for me to merge it as is. I've not tested this btw so am relying on you for testing. Let me know if you think I should also test it.

There is also the other PR by @DaanSelen which is similar to this. Of the two which should be merged/closed?

desertwitch commented:
I'll test this some more today. It's solid code-wise, I'm only concerned the branch feature may be a footgun since the Buildroot manual explicitly warns against it. This could especially be true if it's not updating the branch on rebuilds, and the person doing the builds thinks they are on the latest commit when they're not.

@DaanSelen's version is simple and effective, but relies on a file being downloaded with a separate hash that needs keeping track of (in addition to the commit ID), as opposed to Git taking care of everything. It probably mostly boils down to what you think is more convenient as maintainer who's doing the releases.

DaanSelen commented:
> Does this need further testing/work or are you happy for me to merge it as is. I've not tested this btw so am relying on you for testing. Let me know if you think I should also test it.
> 
> There is also the other PR by @DaanSelen which is similar to this. Of the two which should be merged/closed?

I think kongle's more native to the buildroot suite.

desertwitch commented:
While pulling the initial branch works, it does not seem to update a changed branch afterwards. So the Buildroot manual seems right in that regard, we could just offer "stable commit" (hardcoded) or a custom Git revision (string, let the user decide for themselves what that should be, but remove the hint about a branch being possible)? Otherwise I think there's a slight mismatch of expectation what it should do and what it will do.

So essentially remove the menu entry for the "master branch" and reword the helptext in the Git revision menu entry (omitting anything about branches). That way the user can use a SHA1 of a commit or a tag, all the same.  And if they're adventurous still use the branch as Git revision, just we're not explicitly advertising it. What do you think @Knogle ?

Knogle commented:
> While pulling the initial branch works, it does not seem to update a changed branch afterwards. So the Buildroot manual seems right in that regard, we could just offer "stable commit" (hardcoded) or a custom Git revision (string, let the user decide for themselves what that should be, but remove the hint about a branch being possible)? Otherwise I think there's a slight mismatch of expectation what it should do and what it will do.
> 
> So essentially remove the menu entry for the "master branch" and reword the helptext in the Git revision menu entry (omitting anything about branches). That way the user can use a SHA1 of a commit or a tag, all the same. And if they're adventurous still use the branch as Git revision, just we're not explicitly advertising it. What do you think @Knogle ?

Makes sense for me so far!

Is this okay for you @PartialVolume ? What do you think @DaanSelen ?

desertwitch commented:
Quick mock-up to showcase what I meant: 
<img width="605" height="294" alt="grafik" src="https://github.com/user-attachments/assets/cd515a22-cbd4-4705-a802-c4b20492d0de" />
<img width="681" height="74" alt="grafik" src="https://github.com/user-attachments/assets/7c845043-b988-4783-9dee-3941b680ae6b" />

**Perhaps "stable commit" should instead be the last stable tag hard-coded, so it's obvious what version it is?**

<img width="584" height="295" alt="grafik" src="https://github.com/user-attachments/assets/ed6eae02-70d7-4d4e-89e9-22b3aedbb050" />

Knogle commented:
Like this, or did i miss something? Please take a look.

desertwitch commented:
Looking good, I sent you a PR (tested) with what I would amend, specifically the pre-configure hook is a better choice (and also current state in most recent master, which we might need to rebase this onto at some point pre-merge): https://github.com/Knogle/shredos.x86_64/pull/2

Knogle commented:
> Looking good, I sent you a PR (tested) with what I would amend, specifically the pre-configure hook is a better choice (and also current state in most recent master, which we might need to rebase this onto at some point pre-merge): [Knogle#2](https://github.com/Knogle/shredos.x86_64/pull/2)

Merged! Thanks for your addition! Great idea. @PartialVolume I think now it's ready :)

PartialVolume commented:
@Knogle I'm having issues merging this due to conflicts, which I can't resolve via the github editor.

However, as mentioned a couple of comments ago by @desertwitch you will probably need to rebase or combine the commits into a single commit with an updated commit description of the change that's been made. It will make it more readable and should then resolve the conflicts with nwipe.hash and nwipe.mk that are apparently too complex to resolve via the Github editor.

Knogle commented:
> @Knogle I'm having issues merging this due to conflicts, which I can't resolve via the github editor.
> 
> However, as mentioned a couple of comments ago by @desertwitch you will probably need to rebase or combine the commits into a single commit with an updated commit description of the change that's been made. It will make it more readable and should then resolve the conflicts with nwipe.hash and nwipe.mk that are apparently too complex to resolve via the Github editor.

I'll do so!

Knogle committed:
package/nwipe: switch from static tarball to Git commit (051e1aa)

This commit updates the nwipe package to build from a specific Git commit
instead of using the previously referenced tarball release. The changes
include:

- Setting NWIPE_SITE to the GitHub repository and NWIPE_SITE_METHOD to git.
- Specifying the commit hash (051e1aa) in NWIPE_VERSION.
- Removing or emptying the nwipe.hash file, as we no longer verify a tarball
  and instead rely on the Git checkout.

This approach allows ShredOS/Buildroot to clone the nwipe repository at
commit 051e1aa, apply the banner patch, and build the nwipe binary directly
from Git.

Knogle commented:
I hope i did it the correct way!

PartialVolume commented:
> I hope i did it the correct way!

Looks good 👍
